Docs | Bucket Replication (#1447)
noobaa/noobaa-operator: doc/bucket-replication.md

[NooBaa Operator](../README.md) /

> # Bucket Replication

Reviewer: I would add a definition of the source bucket and the destination/target bucket at the beginning.

> Bucket replication is a NooBaa feature that allows a user to set a replication policy for all or some objects. The goal of replication policies is simple - to define a target for objects to be copied to.

Reviewer: "a target" or a target bucket?

> When applied to a bucketclass, the policy will apply to all future bucket claims that'll utilize the bucketclass.

Reviewer: I feel like you tried to clarify something here, but in this context it is confusing - maybe move it and expand the explanation?

Reviewer: I think it is better (there is a title and the context is clearer).

> In general, a replication policy is a JSON-compliant string which defines an array containing at least one rule  -

Reviewer: > at least one rule

What if the object is empty? (no rules)

>   - Each rule is an object containing a `rule_id`, a `destination bucket`, and an optional `filter` key that contains a `prefix` field.

Reviewer: ```suggestion
  - Each rule is an object containing a `rule_id`, a `destination_bucket`, and an optional `filter` key that contains a `prefix` field.
```

  - When a filter with prefix is provided - only objects keys that match the prefix will be replicated

> The main mechanism behind bucket replication lists all objects across the two buckets, compares the differences, and copies the missing objects from the source to the target bucket. Log-based optimizations utilize AWS S3 server access logging or Azure Monitor to optimize the replication process by copying objects that have been created or modified since the feature was turned on - effectively allowing users to get up to speed with up-to-date recent objects, as the rest replicate in the background with the classic method.

Reviewer: > Log-based optimizations utilize AWS S3 server access logging or Azure Monitor to optimize the replication process by copying objects that have been created or modified since the feature was turned on - effectively allowing users to get up to speed with up-to-date recent objects, as the rest replicate in the background with the classic method.

a long sentence that was not clear to me in its context.

Reviewer: I think it is better, there is a separation between the bucket replication mechanism and the log-based replication.

## Replication Policy Parameters
As stated above, a replication policy is a JSON-compliant array of rules (examples are provided at the bottom of this section)
  - Each rule is an object that contains the following keys:

>     - `rule_id` - which identifies the rule

Reviewer: How do I know what to set rule_id? (general short description?) can I have a couple of `rule_id`s with the same name?

Reviewer: It's better, it gives more details about the field.

    - `destination_bucket` - which dictates the target NooBaa buckets that the objects will be copied to
    - (optional) `{"filter": {"prefix": <>}}` - if the user wishes to filter the objects that are replicated, the value of this field can be set to a prefix string
    - (optional, log-based optimization, see below) `sync_deletions` - can be set to a boolean value to indicate whether deletions should be replicated

>     - (optional, log-based optimization, see below) `sync_versions` - can be set to a boolean value to indicate whether object versions should be replicated

Reviewer: What do I see below? the example? (it is not clear)
Can you refer to a link for more information about `sync_deletions` and `sync_versions`?

Reviewer: It's better.

In addition, when the bucketclass is backed by namespacestores, each policy can be set to optimize replication by utilizing logs (configured and supplied by the user, currently only supports AWS S3 and Azure Blob):
  - <sup><sub>(optional, only supported on namespace buckets)</sub></sup> `log_replication_info` - an object that contains data related to log-based replication optimization -
    - <sup><sub>(necessary on Azure)</sub></sup> `endpoint_type` - this field can be set to an appropriate endpoint type (currently, only AZURE is supported)

>     - <sup><sub>(necessary on AWS)</sub></sup> `{"logs_location": {"logs_bucket": <>}}` - this field should be set to the location of the AWS S3 server access logs

Reviewer: I'm not sure what you tried to do with the "sup sub", but in the browser it is not readable:
![Screenshot 2024-11-20 at 15 39 06](https://github.com/user-attachments/assets/9e0ea6f0-0636-492e-ace8-2d428b8447c0)

## Examples
Note that the example poicies below can also be saved as files and passed to the NooBaa CLI. In that case, it's necessary to omit the outer single quotes.

> ### AWS replication policy with log optimization:

Reviewer: Please add in the example sections an example of a replication policy without any log optimization.

`'{"rules":[{"rule_id":"aws-rule-1", "destination_bucket":"first.bucket", "filter": {"prefix": "a."}}], "log_replication_info": {"logs_location": {"logs_bucket": "logsarehere"}}}'`

### Azure replication policy with log optimization:

`'{"rules":[{"rule_id":"azure-rule-1", "sync_deletions": true, "sync_versions": false, "destination_bucket":"first.bucket"}], "log_replication_info": {"endpoint_type": "AZURE"}}'`

### Namespace bucketclass with replication to first.bucket:

/path/to/json-file.json is the path to a JSON file which defines the replication policy
```shell
noobaa -n app-namespace bucketclass create namespace-bucketclass single bc --resource azure-blob-ns --replication-policy=/path/to/json-file.json
```
```yaml
apiVersion: noobaa.io/v1alpha1
kind: BucketClass
metadata:
  name: bc
  namespace: app-namespace
spec:
  namespacePolicy:
    type: Single
    single: 
      resource: azure-blob-ns
  replicationPolicy: '[{ "rule_id": "rule-1", "destination_bucket": "first.bucket", "filter": {"prefix": "ba"}}]'
```

> 

Reviewer: Please add the json-file.json example as well

[NooBaa Operator](../README.md) /

# Bucket Replication

> Bucket replication is a NooBaa feature that allows a user to set a replication policy for all or some objects. The goal of replication policies is simple - to define a target bucket for objects to be copied to.

Reviewer: Do we protect the user from setting the same bucket as the source and target bucket?
I think it is worth mentioning it.

To utilize bucket replication, we need to first decide what will be our source bucket and what will be our target bucket. The source bucket is the bucket that contains the objects that we want to replicate, and the target bucket is the bucket that will contain the replicated objects. The replication policy is set on the source bucket, and it defines the target bucket(s) and the rules for replication.

In general, a replication policy is a JSON-compliant string which defines an array containing at least one rule  -
  - Each rule is an object containing a `rule_id`, a `destination_bucket`, and an optional `filter` key that contains a `prefix` field.
  - When a filter with prefix is provided - only objects keys that match the prefix will be replicated

> Behind the scenes, bucket replication esstentialy works by comparing object lists. NooBaa lists all objects on the source and target buckets, and checks which objects are missing on the target bucket. It then copies the missing objects from the source to the target bucket (while adhering to any provided rules).

Reviewer: I think you can give more details about the comparison: What happens when an object is modified (same key)? What happens when an object is deleted?

> It is possible to accelerate replication by utilizing logs - at the time of writing, AWS S3 server access logging or Azure Monitor. This mechanism allows NooBaa to copy only objects that have been created or modified since the feature was turned on, while the rest replicate in the background. This allows users to get up to speed with recent objects, while the classic replication mechanism catches up with the rest.

Reviewer: Suggesting to add a new line.
```suggestion
It is possible to accelerate replication by utilizing logs - currently only supports AWS S3 server access logging or Azure Monitor. This mechanism allows NooBaa to copy only objects that have been created or modified since the feature was turned on, while the rest replicate in the background. This allows users to get up to speed with recent objects, while the classic replication mechanism catches up with the rest.
```

    - (Necessary on Azure) `endpoint_type` - this field can be set to an appropriate endpoint type (currently, only AZURE is supported)
    - (Necessary on AWS) `{"logs_location": {"logs_bucket": <>}}` - this field should be set to the location of the AWS S3 server access logs

## Examples
Note that the example poicies below can also be saved as files and passed to the NooBaa CLI. In that case, it's necessary to omit the outer single quotes.
### AWS replication policy:

`'{"rules":[{"rule_id":"aws-rule-1", "destination_bucket":"first.bucket", "filter": {"prefix": "a."}}]}'`

### AWS replication policy with log optimization:

`'{"rules":[{"rule_id":"aws-rule-1", "destination_bucket":"first.bucket", "filter": {"prefix": "a."}}], "log_replication_info": {"logs_location": {"logs_bucket": "logsarehere"}}}'`

### Azure replication policy with log optimization, deletion and version sync:

`'{"rules":[{"rule_id":"azure-rule-1", "sync_deletions": true, "sync_versions": false, "destination_bucket":"first.bucket"}], "log_replication_info": {"endpoint_type": "AZURE"}}'`

### Namespace bucketclass creation with replication to first.bucket:

> 

Reviewer: I'm suggesting to add an opening sentence - 

> There 2 optional ways to set the bucket class creation with a replication policy:
> 1. With the NooBaa CLI - ...
> 2. With a YAML file: - ...

With the NooBaa CLI -

```shell
noobaa -n app-namespace bucketclass create namespace-bucketclass single bc --resource azure-blob-ns --replication-policy=/path/to/json-file.json
```
/path/to/json-file.json is the path to a JSON file which defines the replication policy, e.g. -
```json
{"rules":[{ "rule_id": "rule-1", "destination_bucket": "first.bucket", "filter": {"prefix": "d"}} ]}
```

With a YAML file:
```yaml
apiVersion: noobaa.io/v1alpha1
kind: BucketClass
metadata:
  name: bc
  namespace: app-namespace
spec:
  namespacePolicy:
    type: Single
    single: 
      resource: azure-blob-ns
  replicationPolicy: '[{ "rule_id": "rule-1", "destination_bucket": "first.bucket", "filter": {"prefix": "ba"}}]'
```

### OBC with specific Replication Policy

Applications that require a bucket to have a specific replication policy can create an OBC and add to the claim 
the `spec.additionalConfig.replication-policy` property.

/path/to/json-file.json is the path to a JSON file which defines the replication policy

> Example:

Reviewer: Here as well (if you can add the JSON file, and add a sentence if there are 2 optional ways or that you are showing the result in the YAML file...).
